foxglove/extensions/pid-panel/src/PIDPanel.tsx

```diff
@@ -0,0 +1,341 @@
+  TableCell,
+  TableRow,
+  TableHead,
+  Table,
+  Box,
+  Button,
+  Grid,
+  TextField,
+} from "@mui/material";
+import { ThemeProvider } from "@mui/material/styles";
+import React, { useEffect, useLayoutEffect, useState } from "react";
+import { JSX } from "react/jsx-runtime";
+import { createRoot } from "react-dom/client";
+
+const PID_TOPIC = "/controls/pid_gains";
+const PID_SERVICE = "/controls/set_pid_gains";
+
+type PIDPanelState = {
+  error?: Error;
+  loop: CustomMsgsPidGainConst.LOOP_POSITION | CustomMsgsPidGainConst.LOOP_VELOCITY;
+  editedGains: Record<number, Record<number, number>>; // Indexed by axis and gain type
+  pid: Record<number, Record<number, Record<number, number>>>; // Indexed by loop, axis, and gain type
+};
+
+function PIDPanel({ context }: { context: PanelExtensionContext }): JSX.Element {
+  const [renderDone, setRenderDone] = useState<(() => void) | undefined>();
+
+  // Triple nested object to store current PID gains
+  const initPid = () => {
+    // Initialize gains to -1 to signify that the true values have not been received yet
+    const gains: Record<number, number> = {
+      [CustomMsgsPidGainConst.GAIN_KP]: -1,
+      [CustomMsgsPidGainConst.GAIN_KI]: -1,
+      [CustomMsgsPidGainConst.GAIN_KD]: -1,
+      [CustomMsgsPidGainConst.GAIN_FF]: -1,
+    };
+    const axes: Record<number, Record<number, number>> = {
+      [CustomMsgsPidGainConst.AXIS_X]: { ...gains },
+      [CustomMsgsPidGainConst.AXIS_Y]: { ...gains },
+      [CustomMsgsPidGainConst.AXIS_Z]: { ...gains },
+      [CustomMsgsPidGainConst.AXIS_ROLL]: { ...gains },
+      [CustomMsgsPidGainConst.AXIS_PITCH]: { ...gains },
+      [CustomMsgsPidGainConst.AXIS_YAW]: { ...gains },
+    };
+    const loops: Record<number, Record<number, Record<number, number>>> = {
+      [CustomMsgsPidGainConst.LOOP_POSITION]: { ...axes },
+      [CustomMsgsPidGainConst.LOOP_VELOCITY]: { ...axes },
+    };
+
+    return loops;
+  };
+
+  // Initialize state
+  const [state, setState] = useState<PIDPanelState>({
+    loop: CustomMsgsPidGainConst.LOOP_POSITION,
+    editedGains: {},
+    error: undefined,
+    pid: initPid(),
+  });
+
+  context.subscribe([{ topic: PID_TOPIC }]);
+
+  // Update state when a new PID message is received
+  useLayoutEffect(() => {
+    context.onRender = (renderState: Immutable<RenderState>, done) => {
+      setRenderDone(() => done);
+      if (renderState.currentFrame && renderState.currentFrame.length > 0) {
+        const lastFrame = renderState.currentFrame.at(-1) as MessageEvent<CustomMsgsPidGains>;
+
+        // Loop through received pid_gains and update our PID values
+        const pidGains = lastFrame.message.pid_gains;
+
+        const updatedPid = { ...state.pid }; // Create a copy of the current pid state
+        for (const pidGain of pidGains) {
+          updatedPid[pidGain.loop]![pidGain.axis] = {
+            ...updatedPid[pidGain.loop]![pidGain.axis],
+            [pidGain.gain]: pidGain.value,
+          };
```

Comment:
This can be simplified to `updatedPid[pidGain.loop]![pidGain.axis][pidGain.gain] = pidGain.value`

Reply:
For some reason you need to use the spread operator here. Hung would know more